src/eval.ml

```diff
@@ -187,34 +196,47 @@ and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
                 env.(slot) <- TE_Some(b)
               else
                 let b = lift vars.(pos) in
-                let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+                let xs = Array.map (fun e -> IntMap.find e id_vars) xs in
                 env.(slot) <- TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs b))
         in
         List.iter fn env_builder;
         (* Actually perform the action. *)
         Some(Bindlib.msubst act env, stk)
-    | Cond({ ok ; cond ; fail })                          ->
+    | Cond({ok; cond; fail})                              ->
         let next =
           match cond with
           | CondNL(i, j) ->
               if eq_modulo vars.(i) vars.(j) then ok else fail
-          | CondFV(xs,i) ->
-              let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+          | CondFV(i,xs) ->
+              let allowed =
+                (* Variables that are allowed in the term. *)
+                let fn id =
+                  try IntMap.find id id_vars with Not_found -> assert false
+                in
+                Array.map fn xs
+              in
+              (* Ensure there are no variables from [forbidden] in [b]. *)
+              let no_forbidden b =
+                let forbidden =
```

Comment:
This is independent of b; it should be computed only once:
```
              let no_forbidden =
                 (* Term variables forbidden in the term. *)
                 let forbidden =         
                  IntMap.filter (fun id _ -> not (Array.mem id xs)) id_vars
                in
                fun b -> IntMap.for_all (fun _ x -> not (Bindlib.occur x b)) forbidden
              in
  ```

Reply:
It must be computed once with the term as it is and once, in case of failure with the snf'd term

Reply:
but forbidden does not depend on b... so it should be defined before introducing b. (it's possible that the compiler does this optimization but it's easy to do it by hand)